sync/block_listener.go

```diff
@@ -85,29 +81,43 @@ func (bl *BlockListener) handleBlock(data service.GossipMessage) {
 		bl.Error("got empty message while listening to gossip blocks")
 		return
 	}
-	blk := &types.Block{}
-	err := types.BytesToInterface(data.Bytes(), blk)
+
+	var blk types.Block
+	err := types.BytesToInterface(data.Bytes(), &blk)
 	if err != nil {
-		bl.Error("received invalid block %v", data.Bytes())
+		bl.Error("received invalid block %v", data.Bytes(), err)
 		return
 	}
-	bl.Log.With().Info("got new block", log.Uint64("id", uint64(blk.Id)), log.Int("txs", len(blk.Txs)), log.Int("atxs", len(blk.ATXs)))
+
+	bl.Log.With().Info("got new block", log.Uint64("id", uint64(blk.Id)), log.Int("txs", len(blk.TxIds)), log.Int("atxs", len(blk.ATxIds)))
 	eligible, err := bl.BlockEligible(&blk.BlockHeader)
 	if err != nil {
-		bl.Error("block eligible check failed %v", blk.ID())
+		bl.Error("block %v eligible check failed ", blk.ID())
 		return
 	}
 	if !eligible {
-		bl.Error("block not eligible, %v", blk.ID())
+		bl.Error("block %v not eligible", blk.ID())
 		return
 	}
-	if err := bl.AddBlock(blk); err != nil {
-		bl.Info("Block already received %v", blk.ID())
+
+	associated, txs, atxs, err := bl.syncMissingContent(&blk.MiniBlock)
+	if err != nil {
+		bl.Error("handleBlock %v failed ", blk.ID(), err)
 		return
 	}
-	bl.Info("added block to database %v", blk.ID())
-	data.ReportValidation(NewBlockProtocol)
-	bl.addUnknownToQueue(blk)
+
+	if associated != nil {
+		bl.ProcessAtx(associated)
```

Comment:
IMO this should happen in AddBlockWithTxs so that handling block is done in one singe funciton

Reply:
i agree this thing was really annoying to figure out
but the flow right now dosen't present  a nice way to do it
i rather we deal with it after noam dose the nipst references